File: leetcode-in-python3/solution/71_simplify_path.py
from typing import List
import unittest


class Solution:
    def simplifyPath(self, path: str) -> str:  # startwith`.`
        stack = []
        paths = [i for i in path.split("/") if i != ""]
        for i in paths:
            if i == ".":
                continue
            elif i == "..":
                if len(stack) > 0:
                    stack.pop()
            else:
                stack.append(i)
        return "/" + "/".join(stack)


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ("/home/", "/home"),
            ("/../", "/"),
            ("/", "/"),
            ("/.", "/"),
            ("/.a", "/.a"),
            ("/..a", "/..a"),
            ("/home//foo/", "/home/foo"),
            ("/a/./b/../../c/", "/c"),
            ("/a/../../b/../c//.//", "/c")

        ]
        self.s = Solution()

    def test_solution(self):
        for path, answer in self.test_case:
            ans = self.s.simplifyPath(path)
            self.assertEqual(answer,ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_71_simplify_path.rs
pub struct Solution {}


impl Solution {
    pub fn simplify_path(path: String) -> String {
        let strings: Vec<&str> = path.as_str().split("/").collect();
        let mut stack: Vec<&str> = vec![];
        for i in strings {
            if i == "." || i == "" {
                continue;
            }
            if i == ".." {
                if !stack.is_empty() {
                    stack.pop();
                }
            } else {
                stack.push(i);
            }
        }
        format!("/{}", stack.join("/"))
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            ("/home/", "/home"),
            ("/../", "/"),
            ("/", "/"),
            ("/.", "/"),
            ("/.a", "/.a"),
            ("/..a", "/..a"),
            ("/home//foo/", "/home/foo"),
            ("/a/./b/../../c/", "/c"),
            ("/a/../../b/../c//.//", "/c")
        ];
        for (nums, answer) in test_case {
            let ans = Solution::simplify_path(nums.to_string());
            assert_eq!(ans, answer);
        }
    }
}
